feat: Add `validate_sql_json` endpoint for checking that a given sql query is valid for the chosen database

bearcage opened the pull request:
### CATEGORY

Choose one

- [ ] Bug Fix
- [X] Enhancement (new features, refinement)
- [ ] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [x] Documentation

### SUMMARY
This is the first step toward supporting live as-you-type
SQL validation against databases that support doing it (for
starters just presto).

I'm trying to stick as much to the existing query machinery
as possible to avoid introducing too much new state for handling
this, so I plumbed it through as a db spec parameter and an
option you can set on an individual query.

As of this commit:

- [X] Plumb validation db_spec options
- [X] Add validation_only option and transform to the sql_json view
- [x] Tests for the above on the backend
- [ ] Debounced requestValidationQuery in onSqlChange
- [ ] Check incoming results -- full query vs validation
- [ ] Tests for the above on the frontend
- [ ] Highlight error lines
- [ ] Print error message in alert-danger box above toolbar controls

### TEST PLAN
Tests incoming in another commit on this PR

### ADDITIONAL INFORMATION
- [X] Has associated issue: #6707
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@xtinec @betodealmeida @khtruong

Commit:
Add "validation_only" queries to the backend

This is the first step toward supporting live as-you-type
SQL validation against databases that support doing it (for
starters just presto).

I'm trying to stick as much to the existing query machinery
as possible to avoid introducing too much new state for handling
this, so I plumbed it through as a db spec parameter and an
option you can set on an individual query.

As of this commit:

- [X] Plumb validation db_spec options
- [X] Add validation_only option and transform to the sql_json view
- [ ] Tests for the above on the backend
- [ ] Debounced requestValidationQuery in onSqlChange
- [ ] Check incoming results -- full query vs validation
- [ ] Tests for the above on the frontend
- [ ] Highlight error lines
- [ ] Print error message in alert-danger box above toolbar controls

xtinec commented:
@bearcage looks like there is a lint error we want to fix https://travis-ci.org/apache/incubator-superset/jobs/526725518

DiggidyDave commented:
A couple of high-level points and suggestions for consideration:

1. I'm not a big database guy, but **the term 'validation query'** seems to have an [established meaning](https://www.google.com/search?q=database+validation+query) and I fear this will be unnecessarily overloading a term and possibly causing confusion.  Someone implementing a new type might ask "does my database support validation queries?" and find [this comment on SO](https://stackoverflow.com/a/10684260/3407502) and do the wrong thing.  Consider:

    - renaming it to something less ambiguous and more self-documenting
    - adding more comments to the base class documenting the member property and method

2. Tying this to a database's built in implementation is perhaps limiting. It seem unlikely that many (any?) new databases will supply this functionality, but if there was a configurable "per-dialect" or "per database-type" endpoint exposing an abstract API that we define, than lots of options could be plugged in there (including forwarding through to the provider). Consider:

    - removing the validation support from `BaseEngineSpec` and having a configurable sql validation endpoint map, which by default has one impl for Hive that forwards to the server.  That way we reserve the right to implement the API to make this feature available to other db types going forward. We might say _But we can always add that later!_, and that is true, but if we stub out the interfaces and make the pattern clear and self-documenting now, then the community might pick up the work. The pattern we pick now is the pattern we are going to live with.

john-bodley reviewed (commented):
A couple of general comments:

1. Would you mind adding some unit tests?
2. Superset supports (and encourages) the use of typing (see `CONTRIBUTING.md`). Would you mind adding type hints to the new functions you added? There’s also a note about how to document exceptions (if still relevant).

bearcage commented:
Initially I had this in a separate endpoint, but it seemed like a lot of duplication of effort to pull over all the other conditional modifications we do on queries before executing them (limit, ctas, and templating). Templating in particular was what initially convinced me to move this in to the query execution flow.

That said, I'm not at all used to OSS projects, so I hadn't considered the rhetorical angle. I'll split it back out.

Point well taken on the name. How about `/validate_sql_json` for the endpoint and "SQL Validation" as the name of the feature?

---

Happy to add tests and typing — I have some tests which I'd planned to stack on this commit. I'll squash them instead. Since this portion of the code isn't particularly liberal with types, do y'all prefer py3-only annotations, or comment annotations?

john-bodley commented:
Please use Python 3 only type annotations as Superset only supports Python 3.6+.

bearcage commented on superset/views/core.py:
```diff
@@ -2593,6 +2594,14 @@ def sql_json(self):
+        # apply validation transform last -- after template processing
+        if validate_only:
+            spec = mydb.db_engine_spec
+            if spec.supports_validation_queries:
```
Yep, refactor fatfinger. Fixed.

bearcage commented on superset/db_engine_specs.py:
```diff
@@ -487,6 +488,16 @@ def get_timestamp_column(expression, column_name):
+        If the underlying engine supports it, modify the query sql to request
+        that the database validate the query instead of running it.
+        """
+        raise Exception(
```
I left this in as a defense in depth — right now it's impossible to hit these since we check the conditional in the query path before ever invoking this, but I figured it was either explicitly throw here or return something invalid that might get handed farther up, like a None, moving the 5xx trace farther from the source.

DiggidyDave commented:
> it seemed like a lot of duplication of effort to pull over all the other conditional modifications we do on queries before executing them

I don't think it is much more effort... you can still use all of that. I _think_ it would literally just become:

```
    @classmethod
    def make_validation_query(cls, sql):
        '''
        throws: Exception if no validator type is configured for engine
        '''
        return SQLValidator.get(cls.engine).validate(sql)
```

And that type would just do the last bit of prep that the method currently does, for that impl. The main thing here is this would move it to a configuration option and provide abstractions for impls, rather than make it an intrinsic property of database types.

Commit:
Review comments, phase 1.

Factors out the validation process to a separate class which
does validation inline, rather than passing it through the
existing query flow implicitly.

This is meant to address Dave's feedback requesting that
the validation flow not be explicitly tied to a query transform
since that's uniquely a presto-ism.

Next up in this stack: unit tests.

bearcage commented:
Tagged [WIP] since I need to update unit tests to go with the review changes.

mistercrunch commented on superset/views/core.py:
```diff
@@ -2528,6 +2530,43 @@ def sql_json(self):
                 'Database with id {} is missing.'.format(database_id))
 
+        # Validation request.
+        if validate_only:
```
Why not a different endpoint for validation?

bearcage replied:
No particular reason, apart from an instinct to try to minimize the changeset — I'm happy to pull it out if you'd prefer it that way!

bearcage replied:
Pulled out to a separate endpoint this morning.

Commit:
Split validation off to a separate endpoint

Just to demo it and see if it's cleaner this wak.

Commit:
Add unit tests covering validation

The new validator files are >90% covered, and the coverage
on the core view is back up to about where it was before this
changeset.

bearcage commented:
@DiggidyDave @mistercrunch since you've both mentioned it at one point or another, I pulled validate out to a separate endpoint.

I think this is probably the right call in the long run since the more I think about it the less it makes sense to fuse validation in to the query path. For presto it's quite similar, but for any other validation scheme I can come up with it won't necessarily be. This also avoids conflating validation requests with query requests if you e.g. track 5xx response rate by endpoint.

---

I've also fixed up the unit tests and picked that commit in to this PR for review together.

---

One remaining question on the backend side of this change — what's the right way to expose, from backend-to-frontend, whether or not the frontend should bother trying to call validate_sql_json at all for a given selected db? In other words, how do I expose the db engine config map (or something derived from it) to the frontend so we don't spam the endpoint with calls that will always error out?

Commit:
Fix flake8 lints

`tox -e flake8` is clean now

DiggidyDave commented on superset/views/core.py:
```diff
@@ -2503,6 +2504,61 @@ def stop_query(self):
+            # TODO: factor the Database object out of template rendering
+            #       or provide it as mydb so we can render template params
+            #       without having to also persist a Query ORM object.
+            return json_error_response(
```
I don't think we want to return 500s here. This is an unexceptional condition, it will happen frequently, right? Maybe 404?

What's the flow expected to be? Is it something like this?

1) Client requests validation...
2) Server says "nope I can't do templates yet"
3a) Client says to itself "awww man" and stops requesting validation until the query changes substantially
or...
3b) Client disregards the response and keeps querying at reasonable points? (I sort of favor 3b because it keeps the client simple and gives us a single endpoint to concern ourselves with w.r.t. scale)

bearcage replied:
I think ideally we'd give the client a way to know ahead of time whether a selected db can do validation at all, so we don't have to try the endpoint and trap the error. I'm not entirely sure how to plumb that part, though, so the client code I have right now does 3b — it just keeps on trying pretending nobody ever told it that it'll never work.

DiggidyDave replied:
I agree if we have a clear path to specifically answer the question of whether it is supported then 3a would be ideal.  What about another endpoint to ask that specific question?  `GET /query-validators/presto --> 200 OK or 404 Not Found`

bearcage replied:
I added another layer of indirection to the config object and put the "top half" of it in FEATURE_FLAGS (which are available to the client side as well).

Now the behavior there will be that if no validator is configured at all for the backend, we don't send any validation requests. If a validator is configured, but for some reason doesn't work, we switch to 3b behavior — keep trying it, believing it should be working.

DiggidyDave commented on superset/views/core.py:
```diff
@@ -2503,6 +2504,61 @@ def stop_query(self):
+
+        spec = mydb.db_engine_spec
+        if not spec.engine in SQL_VALIDATORS_BY_ENGINE:
+            return json_error_response(
```
Similarly, **most** databases are going to turn up empty here in at least the near-medium term. This should not return a 500, it is a reasonable, valid response to say "can't do this".

DiggidyDave replied:
Actually, the more I think about it, this might be a straight-up 200 response. Same with the one re templates. Thoughts?

bearcage replied:
IMO 400 bad request is the right return — we're bombing out because the user asked us to do something we explicitly don't do, so we should let them know that.

This is assuming we're able to plumb through the "can I validate on this backend" db property to the client, of course.

DiggidyDave replied:
Yeah, makes sense.  I could see a case for either, but I think 400 is more to the point... 200 would imply we were combining "do you support this feature?" with "is this sql valid?" which would not be a clean API so... I agree with you.

DiggidyDave commented on superset/views/core.py:
```diff
@@ -2503,6 +2504,61 @@ def stop_query(self):
+                f'you have configured the {validator.name} validator '
+                'correctly and that any services it depends on are up. '
+                f'Exception: {e}')
+            return json_error_response(f'{msg}')
```
I would suggest a 400 (bad request) here... if the user sends malformed, unparseable SQL then this is on them, and does not reflect on the health of the service.

bearcage replied:
This condition's raised whenever the validator excepts out for any reason we didn't predict — it could be a bad request, but the most likely reason to end up on this codepath is that the validator itself errored out. For presto that probably means the backing DB 5xx'd us.

On the client side I'm showing this as a warning box in place of the control that normally shows you what the error text is, saying `The server failed to validate your query using ${validator_name}` — just to let you know "hey, the reason you're not seeing errors highlighted isn't that your query is good, it's just that we don't know if it's good or bad"

DiggidyDave replied:
I'm actually pretty concerned about this... this would allow any user to just spam the backend with bad queries and wake up an on-call at 2am.  Can the validator itself make the distinction between an upstream 500 response from the db and a "couldn't parse" problem?

bearcage replied:
I'm not sure what that distinction would look like — here's the mental model I'm using:

The `validator.validate()` interface should never throw, unless it broke. Given that, anything that throws out from it is legitimately a sever error, and should contribute to paging the oncall — if a user is able to form requests that cause the validator to 5xx, that's legitimate breakage.

Phrased another way, if it shows up as a 5xx when someone tries to run a synchronous query and the backing db is toast, shouldn't validations do the same?

DiggidyDave replied:
OK after our in-person chat I'm sold! Thanks 😆 Def add some explanatory comment to that block on `presto_db.py` line 91.

bearcage replied:
Added! Give that a once-over and let me know if that helps to clarify how that part works, since it's definitely not obvious from the code itself.

DiggidyDave replied:
yeah that is much more clear, thanks a ton.

Commit:
Hacky test fix: bind mock validator for pg/mysql

Commit:
Fix pylint lints

Also fix the flake8s that showed up from fixing pylint.

Commit:
Use 400 Bad Request for validation user-error

Consider it user-error and respond with 400 instead of 500
if somebody requests validation with template parameters,
for a backend with no configured validator, or for an
invalid database ID.

These are in contrast to the actual server error reply for
exceptions thrown by the underlying validator implementation,
which still (IMO correctly) returns 500.

Commit:
Move validator config to a feature flag

Commit:
Add docs on configuring validators

Commit:
Fix linter complaints

bearcage committed:
Merge branch 'lyft-release-sp8' into presto/validation/backend

Commit:
Extend DatabaseError catch block and document it

The pyhive presto client yields EXPLAIN (TYPE VALIDATE)
responses as though they were normal queries -- in other
words it doesn't know that errors for that kind of request
aren't exceptional. To work around that, we have to trap
the exceptions it returns and translate the ones that
express problems with the SQL query.

Commit:
fix postgresql test environment

Commit:
Hit pylint with a hammer

It's convinced these are the wrong type. They aren't.

Commit:
More lint fixes

bearcage commented:
@DiggidyDave @mistercrunch @xtinec  I think all the review comments are addressed here — does anybody have any other issues to raise?

DiggidyDave commented:
👍

bearcage committed:
Merge branch 'lyft-release-sp8' into presto/validation/backend
